presto-main/src/main/java/com/facebook/presto/operator/project/PageProjection.java

```diff
@@ -11,24 +11,20 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.facebook.presto.operator;
+package com.facebook.presto.operator.project;
 
-import com.facebook.presto.spi.RecordCursor;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.Page;
 import com.facebook.presto.spi.block.Block;
-import com.facebook.presto.spi.block.BlockBuilder;
 import com.facebook.presto.spi.type.Type;
 
-import java.util.Set;
-
-public interface ProjectionFunction
+public interface PageProjection
 {
     Type getType();
 
-    void project(int position, Block[] blocks, BlockBuilder output);
-
-    void project(RecordCursor cursor, BlockBuilder output);
+    boolean isDeterministic();
 
-    Set<Integer> getInputChannels();
+    InputChannels getInputChannels();
 
-    boolean isDeterministic();
+    Block project(ConnectorSession session, Page page, SelectedPositions selectedPositions);
```

Comment:
This signature requires iffs on `SelectedPositions.isRange` in each `PageProjection` implementation. It would be cleaner for this interface to have two methods:
```
projectSelectedPositions(ConnectorSession session, Page page, int[] selectedPositions)
projectRange(ConnectorSession session, Page page, int offset, int size)
```
then `SelectedPositions` could have `Block applyProjection(ConnectorSession session, Page page)` method. This would separate `PageProjection` from `SelectedPosition` implementation. Also when adding new type of selected positions we would know which classes need to implement it so there are no leftovers.

Alternatively you could also have an `AbstractPageProjection` class that would dispatch calls to:
```
projectSelectedPositions(ConnectorSession session, Page page, int[] selectedPositions)
projectRange(ConnectorSession session, Page page, int offset, int size)
```